db/migrate/20200519102317_add_provider_relationship_permissions.rb

```diff
@@ -0,0 +1,28 @@
+class AddProviderRelationshipPermissions < ActiveRecord::Migration[6.0]
+  def change
+    create_table :provider_relationship_permissions do |t|
+      t.string :type, null: false
+      t.integer :provider_id, null: false
+      t.integer :related_provider_id, null: false
+
+      t.boolean :view_safeguarding_information, default: false, null: false
+
+      t.timestamps
+    end
+
+    add_foreign_key :provider_relationship_permissions, :providers, column: :provider_id
+    add_foreign_key :provider_relationship_permissions, :providers, column: :related_provider_id
+
+    add_index(
+      :provider_relationship_permissions,
+      %i[type provider_id],
+      unique: true,
+    )
+    add_index(
+      :provider_relationship_permissions,
+      %i[provider_id related_provider_id],
```

Comment:
Wouldn't this cause a bit of trouble? If we had provider 1 ratifying courses at provider 2, then provider 2 would define two relationships: the "as ratifying provider" relationship and the "as training provider" relationship. We'd have to make sure we didn't put the same IDs in `provider_id`/`related_provider_id` in each case, which in practice would mean checking for an existing relationship then reversing the positions. 

Would it work to rename these columns as they were (`training_provider_id` and `ratifying_provider_id`) and making a single, three-column unique index of `ratifying_provider_id`, `training_provider_id` and `type`?

Reply:
I would expect that to produce 2 records:

```
<Provider id: 1> # Ratifying provider
<Provider id: 2> # Training provider

# The STI relationships:
<AccreditedBodyPermissions provider_id: 1, related_provider_id: 2, a_permission: false>
<TrainingProviderPermissions provider_id: 2, related_provider_id: 1, a_permission: true>
```

If we condensed these into a single record with specific column names for ratifying and training providers would we then lose the granularity of the permissions on each relationship?

I like the compound index idea and I'm happy to make the column names more specific to the relationship, maybe I'm misunderstanding the use of the `type` here.

Reply:
Ah, I think we agree. I'm proposing we make the role of each provider explicit in the column name. I might well be missing something, though

```
AccreditedBodyPermissions.create!(training_provider: 1, accredited_body: 2) => OK
TrainingProviderPermissions.create!(training_provider: 1, accredited_body: 2) => OK
AccreditedBodyPermissions.create!(training_provider: 1, accredited_body: 2) => 💥
```